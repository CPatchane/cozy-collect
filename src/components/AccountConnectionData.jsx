import React from 'react'
import ClassNames from 'classnames'

import { translate } from 'cozy-ui/react/I18n'

import ReactMarkdownWrapper from 'components/ReactMarkdownWrapper'
import DataItem from 'components/DataItem'
import { getKonnectorMessage } from 'lib/konnectors'
import styles from 'styles/accountConnectionData'

const AccountConnectionData = ({ t, connector }) => {
  const hasDataTypes = !!(connector.data_types && connector.data_types.length)
  const serviceMessage = getKonnectorMessage(t, connector, 'service')
  return (
    <div className={ClassNames(styles['col-account-connection-data'])}>
      {serviceMessage && (
        <div>
          <h4>{t('account.config.data.service.description')}</h4>
          <p>
            <ReactMarkdownWrapper source={serviceMessage} />
          </p>
        </div>
      )}
      <h4>{t('account.config.data.title')}</h4>
      {hasDataTypes && (
        <ul className={styles['col-account-connection-data-access']}>
          {connector.data_types.map((data, index) => (
            <DataItem dataType={data} key={index} />
          ))}
        </ul>
      )}
      {!hasDataTypes && <p>{t('dataType.none', { name: connector.name })}</p>}
    </div>
  )
}

export default translate()(AccountConnectionData)
