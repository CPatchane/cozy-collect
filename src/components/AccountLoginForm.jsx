import React, { PureComponent } from 'react'
import { connect } from 'react-redux'

import { AccountForm } from 'cozy-harvest-lib'
import flag from 'cozy-flags'

import LegacyAccountLoginForm from './LegacyAccountLoginForm'
import { getKonnector } from 'ducks/konnectors'

export class AccountLoginForm extends PureComponent {
  render() {
    const { account, konnector } = this.props
    const initialValues = account ? account.auth : {}
    return flag('harvest') ? (
      <AccountForm initialValues={initialValues} {...konnector} />
    ) : (
      <LegacyAccountLoginForm {...this.props} />
    )
  }
}

const mapStateToProps = (state, ownProps) => ({
  konnector: getKonnector(state.cozy, ownProps.connectorSlug)
})

export default connect(mapStateToProps)(AccountLoginForm)
