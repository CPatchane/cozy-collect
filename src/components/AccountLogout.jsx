import React from 'react'

import { translate } from 'cozy-ui/react/I18n'
import { DeleteAccountButton } from 'cozy-harvest-lib'

import styles from 'styles/accountLogout'

export const AccountLogout = ({ t, account, onError, onSuccess }) => {
  return (
    <div className={styles['col-account-form-delete']}>
      <h4>{t('account.disconnect.title')}</h4>
      <p>{t('account.disconnect.description')}</p>
      <DeleteAccountButton
        account={account}
        onSuccess={onSuccess}
        onError={onError}
        extension="full"
      />
    </div>
  )
}

export default translate()(AccountLogout)
