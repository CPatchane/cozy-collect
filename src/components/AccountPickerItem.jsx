import styles from '../styles/accountPicker'
import classNames from 'classnames'

import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import { translate } from 'cozy-ui/react/I18n'

import { NavLink } from 'react-router-dom'
import Icon from 'cozy-ui/react/Icon'
import Spinner from 'cozy-ui/react/Spinner'

import { getAccount } from '../ducks/accounts'
import { getAccountLogin, getAccountName } from '../lib/helpers'
import { getErrorTitle } from '../lib/konnectors'
import { isInMaintenance } from '../ducks/konnectors'

const MAINTENANCE = 'MAINTENANCE'
const ERRORED = 'ERRORED'
const HAS_UPDATE = 'HAS_UPDATE'
const CONNECTED = 'CONNECTED'

const getStatus = ({ hasError, hasUpdate, isConnected, inMaintenance }) => {
  if (hasUpdate) return HAS_UPDATE
  if (inMaintenance) return MAINTENANCE
  if (hasError) return ERRORED
  if (isConnected) return CONNECTED
}

export class AccountPickerItem extends PureComponent {
  getStatusContent = () => {
    const { t, inMaintenance, connection, hasUpdate } = this.props
    const { error, hasError, isConnected } = connection

    const accountStatus = getStatus({
      hasError,
      hasUpdate,
      isConnected,
      inMaintenance
    })
    switch (accountStatus) {
      case MAINTENANCE:
        return (
          <div className={styles['col-account-picker-button-status']}>
            <span>{t('maintenance.service')}</span>
          </div>
        )
      case ERRORED:
        return (
          <div
            className={classNames(
              styles['col-account-picker-button-status'],
              styles['col-account-picker-button-status--error']
            )}
          >
            <span>
              {getErrorTitle(t, error, key => `connection.error.${key}.title`)}
            </span>
            <Icon icon="warning" />
          </div>
        )
      case HAS_UPDATE:
        return (
          <div className={styles['col-account-picker-button-status']}>
            <span>{t('connector.update')}</span>
          </div>
        )
      case CONNECTED:
        return <Icon icon="check-circleless" color="#2BBA40" />
    }
  }

  render() {
    const { connection, konnectorSlug, account } = this.props
    const { isRunning } = connection
    const accountName = getAccountName(account)
    const accountLogin = getAccountLogin(account)
    const nameAndLoginDiffer = accountName !== accountLogin
    const statusContent = this.getStatusContent()
    return (
      <NavLink
        to={`/connected/${konnectorSlug}/accounts/${account._id}`}
        className={classNames(
          styles['col-account-picker-button'],
          styles['col-account-picker-button-account']
        )}
      >
        <div className={styles['col-account-picker-button-infos']}>
          <span>{accountName}</span>
          {nameAndLoginDiffer && <small>{accountLogin}</small>}
        </div>
        {statusContent}
        {isRunning && <Spinner />}
      </NavLink>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  const { konnectorSlug } = ownProps
  return {
    account: getAccount(
      state.cozy,
      ownProps.connection && ownProps.connection.account
    ),
    inMaintenance: isInMaintenance(konnectorSlug)
  }
}

export default connect(mapStateToProps)(translate()(AccountPickerItem))
