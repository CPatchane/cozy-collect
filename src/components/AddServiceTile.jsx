/* global cozy */
import React, { Component } from 'react'
import Icon from 'cozy-ui/react/Icon'
import Spinner from 'cozy-ui/react/Spinner'

import addServiceIcon from '../assets/icons/icon-plus-add-service.svg'

export class AddServiceTile extends Component {
  constructor(props) {
    super(props)
    this.state = {
      redirecting: false
    }

    this.toggleRedirect = this.toggleRedirect.bind(this)
  }

  async toggleRedirect() {
    if (this.state.redirecting) return // don't toggle twice
    this.setState(() => ({ redirecting: true }))
    try {
      await cozy.client.intents.redirect('io.cozy.apps', { type: 'konnector' })
    } catch (error) {
      console.error(error)
      this.setState({
        redirecting: false
      })
    }
  }

  render() {
    const { label } = this.props
    const { redirecting } = this.state
    return (
      <div
        aria-busy={redirecting}
        className="item-wrapper --add-service"
        onClick={this.toggleRedirect}
      >
        {redirecting ? (
          <Spinner color="grey" size="xlarge" />
        ) : (
          <div className="item-icon">
            <Icon icon={addServiceIcon} />
          </div>
        )}
        <span className="item-title">{label}</span>
      </div>
    )
  }
}

export default AddServiceTile
