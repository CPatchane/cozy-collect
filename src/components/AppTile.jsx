import React, { Component } from 'react'

import { translate } from 'cozy-ui/react/I18n'
import AppIcon from 'cozy-ui/react/AppIcon'

const fetchIcon = cozyClient => async url => {
  let icon
  try {
    const resp = await cozyClient.fetch('GET', url)
    if (!resp.ok)
      throw new Error(`Error while fetching icon ${resp.statusText}: ${url}`)
    icon = await resp.blob()
  } catch (error) {
    throw error
  }
  const mimeType = icon.type
  if (mimeType && mimeType.match(/^image\/.*$/)) {
    return URL.createObjectURL(icon)
  }
  throw new Error(`App icon ${url} is not an image.`)
}

export class AppTile extends Component {
  render() {
    const { app, t } = this.props
    const { client } = this.context
    const displayName = app.name_prefix
      ? `${app.name_prefix} ${app.name}`
      : app.name
    return (
      <a className="item-wrapper" href={app.links && app.links.related}>
        <div className="item-icon">
          <AppIcon
            alt={t('app.logo.alt', { name: displayName })}
            app={app}
            fetchIcon={fetchIcon(client)}
          />
        </div>
        <h3 className="item-title">{displayName}</h3>
      </a>
    )
  }
}

export default translate()(AppTile)
