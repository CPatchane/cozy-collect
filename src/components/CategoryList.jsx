import React from 'react'

import { connect } from 'react-redux'
import { Route, withRouter } from 'react-router-dom'

import { translate } from 'cozy-ui/react/I18n'
import KonnectorList from './KonnectorList'
import ScrollToTopOnMount from './ScrollToTopOnMount'

import { getRegistryKonnectorsByCategory } from '../ducks/registry'
import ConnectionManagement from '../containers/ConnectionManagement'

const CategoryList = ({
  t,
  category,
  categories,
  connectors,
  location,
  wrapper
}) => (
  <div className="content">
    <ScrollToTopOnMount target={wrapper} />
    <div className="col-top-bar">
      <h1 className="col-top-bar-title">
        {!categories.includes(category) || category === 'all'
          ? t('nav.providers')
          : t(`category.${category}`)}
      </h1>
    </div>
    <KonnectorList
      base={`${location.pathname}`}
      konnectors={connectors}
      showVoting
    />
    <Route
      path="/providers/:filter/:konnectorSlug"
      render={props => (
        <ConnectionManagement originPath="/providers/:filter" {...props} />
      )}
    />
  </div>
)

const mapStateToProps = (state, ownProps) => {
  const { filter } = ownProps.match && ownProps.match.params
  return {
    connectors: getRegistryKonnectorsByCategory(
      state.registry,
      filter !== 'all' && filter
    )
  }
}

export default connect(mapStateToProps)(translate()(withRouter(CategoryList)))
