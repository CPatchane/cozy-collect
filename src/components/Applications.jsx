import React, { PureComponent } from 'react'
import { connect } from 'react-redux'

import { Query } from 'cozy-client'
import { translate } from 'cozy-ui/react/I18n'

import AppTile from 'components/AppTile'
import LoadingPlaceholder from 'components/LoadingPlaceholder'
import fillWithGhostItems from 'components/helpers/fillWithGhostItems'
import homeConfig from 'config/collect'
import { receiveApps } from 'ducks/apps'

class LoadingAppTiles extends PureComponent {
  render() {
    const { num } = this.props
    const tiles = []
    for (let i = 0; i < num; i++) {
      tiles.push(
        <div className="item-wrapper">
          <header className="item-header">
            <div className="item-icon">
              <LoadingPlaceholder />
            </div>
          </header>
          <h3 className="item-title">
            <LoadingPlaceholder />
          </h3>
        </div>
      )
    }
    return (
      <div className="list app-list">
        {tiles}
        {fillWithGhostItems(6)}
      </div>
    )
  }
}

export class Applications extends PureComponent {
  render() {
    const { receiveApps } = this.props
    return (
      <div className="app-list-wrapper">
        <Query
          query={client => {
            return client.all('io.cozy.apps')
          }}
        >
          {({ data, fetchStatus }) => {
            if (fetchStatus === 'loaded') {
              receiveApps(data)
            }
            return fetchStatus !== 'loaded' ? (
              <LoadingAppTiles num="3" />
            ) : (
              <div className="list app-list" data-tutorial="home-apps">
                {data
                  .filter(
                    app =>
                      app.state !== 'hidden' &&
                      !homeConfig.filteredApps.includes(app.slug)
                  )
                  .map((app, index) => (
                    <AppTile key={index} app={app} />
                  ))}
                {fillWithGhostItems(6)}
              </div>
            )
          }}
        </Query>
      </div>
    )
  }
}

const mapDispatchToProps = dispatch => {
  return {
    receiveApps: apps => dispatch(receiveApps(apps))
  }
}

export default connect(
  null,
  mapDispatchToProps
)(translate()(Applications))
