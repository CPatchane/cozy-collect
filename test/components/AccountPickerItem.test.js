'use strict'

/* eslint-env jest */

import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-15'

import { tMock } from '../jestLib/I18n'
import { AccountPickerItem } from '../../src/components/AccountPickerItem'

configure({ adapter: new Adapter() })

const mockAccount = {
  _id: '123456789',
  accountName: 'wawawa youpiyo youpiyai',
  login: 'wawawa@youpiyo.fr'
}

describe('AccountPickerItem component', () => {
  it('should render correctly if success', () => {
    const mockProps = {
      t: tMock,
      hasUpdate: false,
      konnectorSlug: 'bouilligue',
      account: mockAccount,
      connection: { isConnected: true }
    }
    const component = shallow(<AccountPickerItem {...mockProps} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render a spinner if it is running', () => {
    const mockProps = {
      t: tMock,
      hasUpdate: false,
      konnectorSlug: 'bouilligue',
      account: mockAccount,
      connection: { isRunning: true }
    }
    const component = shallow(<AccountPickerItem {...mockProps} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should display the update message if has update and no errors', () => {
    const mockProps = {
      t: tMock,
      hasUpdate: true,
      konnectorSlug: 'bouilligue',
      account: mockAccount,
      connection: {}
    }
    const component = shallow(<AccountPickerItem {...mockProps} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should display the update message if has update even if is connected', () => {
    const mockProps = {
      t: tMock,
      hasUpdate: true,
      konnectorSlug: 'bouilligue',
      account: mockAccount,
      connection: { isConnected: true }
    }
    const component = shallow(<AccountPickerItem {...mockProps} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should display the error if it has error', () => {
    const mockProps = {
      t: tMock,
      hasUpdate: false,
      konnectorSlug: 'bouilligue',
      account: mockAccount,
      connection: { hasError: true, error: new Error('Expected test error') }
    }
    const component = shallow(<AccountPickerItem {...mockProps} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should display the error if it has error even if it has update', () => {
    const mockProps = {
      t: tMock,
      hasUpdate: true,
      konnectorSlug: 'bouilligue',
      account: mockAccount,
      connection: {
        hasError: true,
        error: new Error('Expected test error')
      }
    }
    const component = shallow(<AccountPickerItem {...mockProps} />).getElement()
    expect(component).toMatchSnapshot()
  })
})
