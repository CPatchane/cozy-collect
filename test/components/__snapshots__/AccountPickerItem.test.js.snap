// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountPickerItem component should display the error if it has error 1`] = `
<NavLink
  activeClassName="active"
  aria-current="page"
  className="col-account-picker-button col-account-picker-button-account"
  to="/connected/bouilligue/accounts/123456789"
>
  <div
    className="col-account-picker-button-infos"
  >
    <span>
      123456789
    </span>
    <small />
  </div>
  <div
    className="col-account-picker-button-status col-account-picker-button-status--error"
  >
    <span>
      Expected test error
    </span>
    <Icon
      icon="warning"
      spin={false}
    />
  </div>
</NavLink>
`;

exports[`AccountPickerItem component should display the error if it has error even if it has update 1`] = `
<NavLink
  activeClassName="active"
  aria-current="page"
  className="col-account-picker-button col-account-picker-button-account"
  to="/connected/bouilligue/accounts/123456789"
>
  <div
    className="col-account-picker-button-infos"
  >
    <span>
      123456789
    </span>
    <small />
  </div>
  <div
    className="col-account-picker-button-status col-account-picker-button-status--error"
  >
    <span>
      Expected test error
    </span>
    <Icon
      icon="warning"
      spin={false}
    />
  </div>
</NavLink>
`;

exports[`AccountPickerItem component should display the update message if has update and no errors 1`] = `
<NavLink
  activeClassName="active"
  aria-current="page"
  className="col-account-picker-button col-account-picker-button-account"
  to="/connected/bouilligue/accounts/123456789"
>
  <div
    className="col-account-picker-button-infos"
  >
    <span>
      123456789
    </span>
    <small />
  </div>
  <div
    className="col-account-picker-button-status"
  >
    <span>
      Update available
    </span>
  </div>
</NavLink>
`;

exports[`AccountPickerItem component should display the update message if has update even if is connected 1`] = `
<NavLink
  activeClassName="active"
  aria-current="page"
  className="col-account-picker-button col-account-picker-button-account"
  to="/connected/bouilligue/accounts/123456789"
>
  <div
    className="col-account-picker-button-infos"
  >
    <span>
      123456789
    </span>
    <small />
  </div>
  <div
    className="col-account-picker-button-status"
  >
    <span>
      Update available
    </span>
  </div>
</NavLink>
`;

exports[`AccountPickerItem component should render a spinner if it is running 1`] = `
<NavLink
  activeClassName="active"
  aria-current="page"
  className="col-account-picker-button col-account-picker-button-account"
  to="/connected/bouilligue/accounts/123456789"
>
  <div
    className="col-account-picker-button-infos"
  >
    <span>
      123456789
    </span>
    <small />
  </div>
  <Wrapper />
</NavLink>
`;

exports[`AccountPickerItem component should render correctly if success 1`] = `
<NavLink
  activeClassName="active"
  aria-current="page"
  className="col-account-picker-button col-account-picker-button-account"
  to="/connected/bouilligue/accounts/123456789"
>
  <div
    className="col-account-picker-button-infos"
  >
    <span>
      123456789
    </span>
    <small />
  </div>
  <Icon
    color="#2BBA40"
    icon="check-circleless"
    spin={false}
  />
</NavLink>
`;
